### ISSUE TITLE
Error message when using texture from viewport inside itself

### ISSUE THREAD

**Issue by JonqsGames** [godotengine/godot#81928]
### Godot version

4.2

### System information

Windows 11

### Issue description

Implementing a similar effect than describe in : https://www.youtube.com/watch?v=VSwVwIYEypY
I think it's intended but server spams an error :
`
RenderingDeviceVulkan::draw_list_bind_uniform_set: Attempted to use the same texture in framebuffer attachment and a uniform (set: 1, binding: 3), this is not allowed.
`
It's still works as expected, could this error be removed or ignored ?
Maybe it could be a warning clarifying that this has 1 frame delay.

*Maybe doing this is not recommended in that case this issue can be usefull to other people know why*

### Steps to reproduce

Project from the tutorial is raising the error : https://github.com/CBerry22/Godot-Water-Ripple-Simulation-Shader

### Minimal reproduction project

Create a viewport
Inside that viewport use the ViewportTexture from parent in a shader.

**Comment by nicholaskim6** [comment]
I'm also seeing this constantly and am wondering if there's a fix

**Comment by clayjohn** [comment]
I think some context behind this error would be helpful. 

First, this error message has been around since the Vulkan renderer was written. Its not a new error.
Second, this error is for your information, it doesn't change any behaviour and its only enabled in debug builds
Last, this error protects you from doing something that is not technically supported by OpenGL or Vulkan. 

That last point needs some clarification, especially since it appears to work when using OpenGL and Vulkan. 

Writing to a framebuffer (viewport) and sampling that framebuffer at the same time is undefined behaviour in OpenGL and Vulkan. Undefined behaviour means that the specification doesn't tell GPU driver authors what to do. In Vulkan, you have the concept of framebuffer local load operations which allow you to read from the current pixel in the framebuffer, but this has some clear restrictions that make it very limited.

That being said, if a GPU driver author wants to allow you to read from the same framebuffer you are writing to, they can allow it. You just _can't count on it working on other machines_. The catch is that there is no way to check in advance if the GPU driver supports this. Its totally random, and they are not guaranteed to maintain support across driver updates either. So it might work on your computer one day, then break the next when you get an update. 

In my experience, desktop GPU drivers mostly allow reading from the currently bound framebuffer as they tend to not be very strict while mobile drivers tend to forbid it entirely. Notably, ANGLE also totally forbids reading from the currently bound framebuffer. So any shader that relies on doing so will fail when exported to the web. 

If you are writing a game that will only ever run on your device (and you control driver updates), its fine to rely on this behaviour. Otherwise, it is much safer to never rely on it. Since the Godot team doesn't control the hardware that Godot runs on, we need to forbid reading from the currently bound framebuffer (which is what ANGLE does). 

If anything, the error should be added to the compatibility renderer as well so that users can get a helpful error before they rely on undefined behaviour that will break their game on mobile and web and any number of other devices

**Comment by inkusgames** [comment]
Thank you for the details breakdown of the error message. 

Is there an alternative and less risky approach if we are trying to write a processing shader the uses its output on the next pass as a source?

**Comment by clayjohn** [comment]
> Thank you for the details breakdown of the error message.
> 
> Is there an alternative and less risky approach if we are trying to write a processing shader the uses its output on the next pass as a source?

Yep, you just need to use two viewports. The reproduction project from the OP was actually updated to use the correct approach 6 days after this PR was opened https://github.com/CBerry22/Godot-Water-Ripple-Simulation-Shader. You can reference that project both before and after the change to see exactly what you need to do

**Comment by inkusgames** [comment]
I reworked the MRP to correctly implement the solution as @clayjohn pointed out.

https://github.com/inkusgames/godot4_shader_viewport_example

Thanks for the knowledge.
